tests/allotrope/schema_parser/schema_cleaner_test.py

```diff
@@ -0,0 +1,1392 @@
+                    },
+                    "required": ["childKey2", "parentKey"],
+                },
+                {
+                    "properties": {
+                        "parentKey": "value",
+                        "childKey1": "value",
+                    },
+                    "required": ["parentKey"],
+                },
+            ]
+        },
+    )
+
+
+def test_combine_anyof_with_parent_anyof_required_keys() -> None:
+    schema = {
+        "properties": {
+            "key1": "value",
+            "key2": "otherValue",
+        },
+        "anyOf": [
+            {"required": ["key1"]},
+            {"required": ["key2"]},
+        ],
+    }
+    validate_cleaned_schema(
+        schema,
+        {
+            "anyOf": [
+                {
+                    "properties": {"key1": "value", "key2": "otherValue"},
+                    "required": ["key1", "key2"],
+                },
+                {
+                    "properties": {"key1": "value", "key2": "otherValue"},
+                    "required": ["key1"],
+                },
+                {
+                    "properties": {"key1": "value", "key2": "otherValue"},
+                    "required": ["key2"],
+                },
+            ]
+        },
+    )
+
+
+def test_fixes_singlular_allof() -> None:
+    schema = {
+        "allOf": [{"key": "value"}],
+    }
+    validate_cleaned_schema(schema, {"key": "value"})
+
+
+def test_fixes_oneof_nested_in_allof() -> None:
+    schema = {
+        "allOf": [
+            {"properties": {"key1": "value1"}},
+            {
+                "oneOf": [
+                    {"properties": {"key2": "value2"}},
+                    {"properties": {"key3": "value3"}},
+                ]
+            },
+        ]
+    }
+    expected = {
+        "oneOf": [
+            {
+                "allOf": [
+                    {"properties": {"key1": "value1"}},
+                    {"properties": {"key2": "value2"}},
+                ]
+            },
+            {
+                "allOf": [
+                    {"properties": {"key1": "value1"}},
+                    {"properties": {"key3": "value3"}},
+                ]
+            },
+        ]
+    }
+    validate_cleaned_schema(schema, expected)
+
+    # Flip so oneOf is first in allOf list, to cover all branches.
+    schema["allOf"] = schema["allOf"][::-1]
+    expected["oneOf"] = [{"allOf": value["allOf"][::-1]} for value in expected["oneOf"]]
+    validate_cleaned_schema(schema, expected)
+
+
+def test_fixes_oneof_nested_in_allof_in_reference() -> None:
+    schema = {
+        "allOf": [{"properties": {"key1": "value1"}}, {"$ref": "#/$defs/oneOfSchema"}],
+        "$defs": {
+            "oneOfSchema": {
+                "oneOf": [
+                    {"properties": {"key2": "value2"}},
+                    {"properties": {"key3": "value3"}},
+                ]
+            }
+        },
+    }
+    expected = {
```

Comment:
another understanding check: why don't each of the nested `allOf`s get combined, like what happens in `test_combine_allof`? (I guess this is a general question I have about all the tests, why these nestings don't all get collapsed all the way down? or do the tests only test for what happens after one iteration of cleaning?)

Reply:
Good question - the answer is that we quit when we know the schema generation script can handle it :D 

See the end of `_combine_allof`

My thinking here was to do as little editing of the schema as I could get away with, in case we have to do more work in the future.

Luckily, the way the code works, no one will ever see these confusing schemas, except the poor souls who edit this code or review my PRs 😭

Reply:
ah... "as little as possible" makes sense and was definitely the key piece I was missing 🤦‍♂️ thank you for clarifying!